### DISCUSSION TITLE
Method to obtain vector within a 'MechanicalObject' object

### DISCUSSION THREAD

**Discussion by jemmyjiang15** [sofa-framework/sofa#3779] [Write / Run a simulation] [answered]
Hi,

I am currently using one of the Python examples provided with SOFA (SoftFingerDesign) and extending that to simulate some soft fingers for design optimisation. 

![image](https://user-images.githubusercontent.com/25949334/231063734-2141c8cf-b8c9-462e-b516-21ba737e21e9.png)

My current problem is, within the MechanicalObject object, there is a particular value that I would like to access the data from, namely "contraint_dx(V_DERIV)".

![image](https://user-images.githubusercontent.com/25949334/231063924-08ed3ee3-9b6e-49dd-aa5b-c7a1212fcbbd.png)

However, I have tried many different combinations of ways to access that data, including `SelfCollisionMesh1.MechanicalObject.constraint_dx.value`. This method works with other vectors within MechanicalObject, such as the  'position' vector (through `SelfCollisionMesh1.MechanicalObject.position.value`), but with all the variable names I have tried, it only returns with a key error, meaning that I likely have used the incorrect variable name.

Could I get some help with a method on how to obtain this vector during simulation?

Thanks,
Jemmy

**Comment by alxbilger** [comment]
Hi,

this Data is created at runtime by the constraint solver. It does not exist at the initialization step of the simulation. Could you try to access it, but only after the first time step?

**Reply by jemmyjiang15** [reply]
The code I have written to obtain the `constraint_dx` data is under an `onAnimateBeginEvent()` function based on a `Sofa.Core.Controller` class object, which is called at every time step interval. 

![image](https://user-images.githubusercontent.com/25949334/231306169-e3b12203-eebd-4cac-b1ef-4eb5843beeb3.png)

I'm receiving this error in cmd at every time step after I begin to simulate so I'm assuming the data still cannot be accessed even after the first timestep. I'm wondering if I have the correct variable to be able to extract the `constraint_dx` vector data?

**Reply by alxbilger** [reply]
You could use the following Python function on the MechanicalObject to debug your problem:

- `getDataFields`: vector containing all the fields of this object
- `getData`: Get the data field given its name
- `findData`: Find a data field given its name. Return NULL if not found

See if `constraint_dx` appears when calling `getDataFields` (but I guess it won't).

`constraint_dx` is dynamically added to the MechanicalObject. But I don't see how it makes a difference for a read access...

**Reply by damienmarchal** [reply]
Hi, 

To get the list of the data in a node I did:
```python
list(myMechanicalObject.__data__)
```
which returns
```console
['name', 'printLog', 'tags', 'bbox', 'componentState', 'listening', 'position', 'velocity', 'force', 'rest_position', 'externalForce', 'derivX', 'free_position', 'free_velocity', 'constraint', 'mappingJacobian', 'reset_position', 'reset_velocity', 'restScale', 'useTopology', 'showObject', 'showObjectScale', 'showIndices', 'showIndicesScale', 'showVectors', 'showVectorsScale', 'drawMode', 'showColor', 'translation', 'rotation', 'scale3d', 'translation2', 'rotation2', 'size', 'reserve', 'lambda(V_DERIV)', 'constraint_dx(V_DERIV)']
```
We can see that constraint_dx is there but, but its real name is "constraint_dx(V_DERIV)". This name, with the parenthesis is not supported using the pythonic API (the parenthesis conflict with the python syntax). So we must use getData("constraint_dx(V_DERIV)").  


To make things more complex, this data is added at runtime... this means that the scene has to be properly initialized. Initialization is not done at the scene creation from python. 

Ex:
```python
import Sofa
n = Sofa.Core.Node("test")
n.addObject("RequiredPlugin", name="Sofa.Component.Constraint.Lagrangian.Solver")
mm = n.addObject("MechanicalObject")
gcc = n.addObject("GenericConstraintSolver")

# at that print there is not constraint_dx
print(list(mm.__data__))

# manually force initialization of the scene 
n.init() 

# so now the constraint_dx is registered and appears. 
print(list(mm.__data__))

# but if we do ... 
print(mm.getData("constraint_dx(V_DERIV"))
```
It returns None...because there is a bug in how MechanicalObject register the name in sofa (there is an alias mechanism caching the name).

**Reply by damienmarchal** [reply, marked as answer]
After some investigation, there is two problem in this discussion.
- the dynamic data are not registered properly using Base::addAlias so they cannot be queried using getData, and the corresponding python binding.
- the naming scheme with the type in parenthesis is misleading:  "constraint_dx(V_DERIV)" -> "constraint_dx" 

I provided a preliminary fix in #3783. Deprecating the parenthesis based naming scheme may be breaking some other behavior so maybe I will have to fallback just fixing the addAlias registration.

**Reply by jemmyjiang15** [reply]
Thanks for explaining the issue, so this means that there is no way to read the constraint_dx vector variable through code in the current version of SOFA I have?

**Comment by damienmarchal** [comment]
@alxbilger, actually the way vReallocImpl and write() are called is very sofa-esque :)

EDIT 1: vReallocImpl(), calls write(), which in turn allocate the data if not present() using hardcoded code path :)
EDIT 2: now I'm looking at vAllocImpl() vs vReallocImpl() and feel so sad.
